### PR TITLE
Add tests against multiple MongoDB versions

### LINKED ISSUE #7160: Add tests against multiple MongoDB versions
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation
Parse Server CI currently only tests against MongoDB 3.6.21 and 4.0.4.

- This does not reflect the variety of MongoDB versions in use in real word deployments.
- Parse Server may not be fully compatible with the MongoDB Atlas free tier anymore, which runs MongoDB 4.4.3. This means that the docs that describe how to quickly get started using MongoDB Atlas are not practical anymore for developers who just want to try out Parse Server, because they can only choose a paid tier.

### Feature / Enhancement Description
Keep the MongoDB versions for CI tests in sync with the officially supported MongoDB versions.

That means:
- Removing the CI tests for MongoDB 3.6 (EOL date April 2021) and instead test for the following 3 versions:

  | Version     | EOL Date     | Latest Patch Release |
  |-------------|--------------| - |
  | ~~MongoDB 3.6~~ | ~~April 2021~~   |~~3.6.21~~ |
  | MongoDB 4.0 | January 2022 | 4.0.22 |
  | MongoDB 4.2 | TBD          | 4.2.12 |
  | MongoDB 4.4 | TBD          | 4.4.3 |

- When a new patch version is released, update the CI test version as well.

In addition, the Readme should contain a MongoDB compatibility table to show against which MongoDB versions which version of Parse Server is tested against.

### Example Use Case
(none)

### Alternatives / Workarounds
(none)

### 3rd Party References
(none)

### PR DESCRIPTION
**Pull request by mtrezza**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
Add tests for multiple MongoDB versions.

Related issue: closes #7160

### Approach
- Added MongoDB versions in CI matrix.
- Added spec helper `it_only_mongodb_version(version)` that only executes a test if the specified semver version satisfies the MongoDB version; allows to write duplicates tests for MongoDB versions and easily discard them by their semver notation once testing for a specific MongoDB version is removed from the CI.
- Added npm scripts to easily run tests against the relevant MongoDB versions that have to pass in the CI, e.g. `npm run test:mongodb:4.4.3`; or simply run against any version the developer prefers with `npm run test:mongodb --dbversion=[VERSION]`.

### TODOs before merging

- [x] Ensure the latests patch of each officially supported MongoDB version is tested in CI
- [x] Ensure CI tests pass
- [x] Add MongoDB compatibility table to README.md
- [x] Add entry to changelog

### CODE REVIEW

**Comment by dplewis** [comment]
We did run into an issue when we where trying to migrate to Github Actions. We use 4.0.4 of mongo since it was the only version that works with the tests. You can view the conversation (trial and error) [here](https://github.com/parse-community/parse-server/pull/7035#issuecomment-737750535)

**Comment by mtrezza** [comment]
@dplewis All test pass now, they did not pass because:
- some were poorly written and did not follow MongoDB specification, MongoDB 4.4 is more strict about that
- some had to be adapted to different behavior in different versions of MongoDB

In fact, Parse Server was not fully compatible since MongoDB 4.2 (2019).

To make sure this doesn't happen again, I added a compatibility table to the README. We should keep an eye on the [MongoDB support policy](https://www.mongodb.com/support-policy) and update the MongoDB versions we test against.

TODOs after merging this PR:
- In April 2021 MongoDB 3.6 reaches its end of life support. We can then remove it from the CI tests to focus on the newer versions.
- Regularly update the patch versions for all MongoDB minor versions; MongoDB still releases patches for versions <4.4, see compatibility table.

File: README.md
Comment (dplewis) [inline README.md:50]: * Compatibility

File: spec/.eslintrc.json
Comment (dplewis) [inline spec/.eslintrc.json:22]: Should we update the contributors guide?

**Review by dplewis — CHANGES_REQUESTED** [review]
Good work getting this in! A quick nit and a suggestion.

File: README.md
Comment (mtrezza) [reply README.md:50]: I think I finally need a spell checker.

Comment (mtrezza) [reply README.md]: done

File: spec/.eslintrc.json
Comment (mtrezza) [reply spec/.eslintrc.json:22]: done

**Comment by mtrezza** [comment]
The tests are now covering all relevant minor MongoDB versions:

- Mongo 4.4.3, ReplicaSet, WiredTiger
- Mongo 4.2.12, ReplicaSet, WiredTiger
- Mongo 4.0.22, ReplicaSet, WiredTiger (MongoDB 4.0.x released in June 2018)
- Mongo 3.6.21, Standalone, MMAPv1 (can be removed in April 2021 which is the end of life support date)

That means it covers MongoDB 4.x in its most common set-up (ReplicaSet, WT) and MongoDB 3.6 as a Standalone with the MMAPv1 storage engine (which has been deprecated in MongoDB 4.2).

**Review by dplewis — APPROVED** [review]

**Comment by mtrezza** [comment]
Side note, the fact that this adds two additional test environments obviously increases the chance of running into a flaky test. This should force us to focus on these and fix them, rather than dragging them along. Most of the flakiness I have fixed in the past came from poor test concepts (race conditions, global variable overwriting, etc). This PR will scrutinize new tests more thoroughly and therefore should mitigate the introduction of new flaky tests.

**Comment by dplewis** [comment]
@mtrezza I agree the test suite is in need of some love. This PR will also help me with the schema improvements that I've been trying to add for a while now. You inspired me to clean up the JS SDK.

**Review by dplewis — APPROVED** [review]
LGTM!

**Comment by dplewis** [comment]
@davimacedo Anything you want to add?

**Comment by mtrezza** [comment]
Do not merge this yet, I am working on a CI script that automatically checks for newer MongoDB versions.

**Comment by mtrezza** [comment]
I added a CI self check job that automatically checks the CI workflow whether:
- The environments use the latest patch version of MongoDB (e.g. fails when using MongoDB 4.3.1 but there is a release 4.3.2 fails)
- There are a higher minor or major MongoDB version available (e.g. fails when using MongoDB 4.3.x, but there is a release 4.4.x or 5.x); ignores all pre-release versions such a release candidates or alpha / beta versions.



Example:


image

Fails because there is newer MongoDB patch version.

- Is it possible to manually exclude certain MongoDB releases. It currently excludes MongoDB 4.7.x because this is a development release (not obvious, but I confirmed this with MongoDB).

**Comment by mtrezza** [comment]
If there is no objection, I will go ahead and merge this later today.

**Review by davimacedo — APPROVED** [review]

**Comment by mtrezza** [comment]
I noticed that the Parse Server repo only has Snyk as required passing test. Shouldn't these MongoDB be required as well? @davimacedo @dplewis

**Comment by davimacedo** [comment]
Where do you see that @mtrezza ?

**Comment by mtrezza** [comment]
In the repo Settings > Branches > Branch Protection Rules > Edit > Require status checks to pass before merging.

**Comment by davimacedo** [comment]
Yes. I've just seem it here. I believe we should include these checks. @dplewis thoughts?

**Comment by dplewis** [comment]
I agree, I never noticed out of habit.

**Comment by davimacedo** [comment]
I've just activated. Let's see how it works.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)

**Pull request merged**
